### PR TITLE
[Merged by Bors] - feat(Data/Setoid): another way to define piecewise functions

### PR DESCRIPTION
**Pull request by shuxuezhuyi**
Use `IndexedPartition` to define `piecewise` functions. This is natural for piecewise functions that are split into many pieces.​

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Setoid/Partition.lean [hunk 1]
```diff
@@ -475,4 +476,79 @@ theorem proj_fiber (x : hs.Quotient) : hs.proj ⁻¹' {x} = s (hs.equivQuotient.
...
+    rw [hjy]
+    simp only [mem_image]
+    use y
+    simp only [and_true]
+    exact mem_index hs y⟩
+  have hij : i = j := hik.trans hjk.symm
+  rw [hix, hjy, hij] at hfxy
+  refine (h_injOn j) ?_ (mem_index hs y) hfxy
+  rw [← hij]
+  exact mem_index hs x
```
Comment (winstonyin) [inline Mathlib/Data/Setoid/Partition.lean:525]: Golfed a bit
```suggestion
  intro x y hfxy
  let i := index hs x
  have hix : piecewise hs f x = f i x := rfl
  let j := index hs y
  have hjy : piecewise hs f y = f j y := rfl
  have hz : piecewise hs f x ∈ (⋃ i' ∈ univ, (f i') '' (s i')) := by
    simp only [mem_univ, iUnion_true, mem_iUnion, mem_image]
    exact ⟨i, x, mem_index hs x, rfl⟩
  choose k hk_exist hk_unique using pairwiseDisjoint_unique h_disjoint hz
  set g := piecewise hs f
  have hik : i = k := hk_unique i ⟨trivial, ⟨x, mem_index hs x, rfl⟩⟩
  have hjk : j = k := hk_unique j ⟨trivial, ⟨y, mem_index hs y, by rw [hfxy, hjy]⟩⟩
  have hij : i = j := hik.trans hjk.symm
  rw [hix, hjy, hij] at hfxy
  refine (h_injOn j) ?_ (mem_index hs y) hfxy
  rw [← hij]
  exact mem_index hs x
```

File: Mathlib/Data/Setoid/Partition.lean [hunk 2]
```diff
@@ -475,4 +476,79 @@ theorem proj_fiber (x : hs.Quotient) : hs.proj ⁻¹' {x} = s (hs.equivQuotient.
...
+    refine injOn_of_injective ?_ (⋃ (i : ι), s i)
+    refine piecewise_inj hs ?h_injOn ?h_disjoint
+    · exact fun i ↦ BijOn.injOn (hf i)
+    · unfold PairwiseDisjoint
+      simp only [fun i => BijOn.image_eq (hf i)]
+      intro i _ j _ hij
+      exact ht.disjoint hij
+  have hg := bijOn_iUnion hg_bij hg_inj
+  rw [hs.iUnion, ht.iUnion] at hg
+  exact bijective_iff_bijOn_univ.mpr hg
```
Comment (winstonyin) [inline Mathlib/Data/Setoid/Partition.lean:552]:
```suggestion
  set g := piecewise hs f with hg
  have hg_bij : ∀ i, BijOn g (s i) (t i) := by
    intro i
    refine BijOn.congr (hf i) ?_
    intro x hx
    rw [hg, piecewise, hs.mem_iff_index_eq.mp hx]
  have hg_inj : InjOn g (⋃ i, s i) := by
    refine injOn_of_injective ?_ (⋃ (i : ι), s i)
    refine piecewise_inj hs (fun i ↦ BijOn.injOn (hf i)) ?h_disjoint
    simp only [fun i ↦ BijOn.image_eq (hf i)]
    intro i _ j _ hij
    exact ht.disjoint hij
  rw [bijective_iff_bijOn_univ, ← hs.iUnion, ← ht.iUnion]
  exact bijOn_iUnion hg_bij hg_inj
```

**Comment by winstonyin** [comment]
I made some suggestions to clean up / shorten the proofs. @j-loreaux did you have more comments in mind?

**Comment by j-loreaux** [comment]
I don't remember why I self-requested a review here.

File: Mathlib/Data/Setoid/Partition.lean [hunk 3]
```diff
@@ -502,4 +503,58 @@ theorem proj_fiber (x : hs.Quotient) : hs.proj ⁻¹' {x} = s (hs.equivQuotient.
     exact Quotient.eq''
 #align indexed_partition.proj_fiber IndexedPartition.proj_fiber
 
+/-- Combine functions with disjoint domains into a new function.
+You can use the regular expression `def.*piecewise` to search for
+other ways to define piecewise functions in mathlib4.-/
+def piecewise {β: Type*} (f : ι → α → β) : α → β := fun x => f (hs.index x) x
```
Comment (jcommelin) [inline Mathlib/Data/Setoid/Partition.lean:509]:
```suggestion
def piecewise {β : Type*} (f : ι → α → β) : α → β := fun x => f (hs.index x) x

lemma piecewise_apply {β : Type*} {f : ι → α → β} (x : α) : hs.piecewise f x = f (hs.index x) x :=
  rfl
```

File: Mathlib/Data/Setoid/Partition.lean [hunk 4]
```diff
@@ -502,4 +503,58 @@ theorem proj_fiber (x : hs.Quotient) : hs.proj ⁻¹' {x} = s (hs.equivQuotient.
...
+
+open Function
+
+/-- A family of injective functions with pairwise disjoint
+domains and pairwise disjoint ranges can be glued together
+to form an injective function.-/
+theorem piecewise_inj {β : Type*} {f : ι → α → β}
+    (h_injOn : ∀ i, InjOn (f i) (s i))
+    (h_disjoint : PairwiseDisjoint (univ : Set ι) fun i => (f i) '' (s i)):
```
Comment (jcommelin) [inline Mathlib/Data/Setoid/Partition.lean:518]:
```suggestion
    (h_disjoint : PairwiseDisjoint (univ : Set ι) fun i => (f i) '' (s i)) :
```

File: Mathlib/Data/Setoid/Partition.lean [hunk 5]
```diff
@@ -502,4 +503,58 @@ theorem proj_fiber (x : hs.Quotient) : hs.proj ⁻¹' {x} = s (hs.equivQuotient.
...
+    exact ⟨i, x, mem_index hs x, rfl⟩
+  choose k hk_exist hk_unique using pairwiseDisjoint_unique h_disjoint hz
+  set g := piecewise hs f
+  have hik : i = k := hk_unique i ⟨trivial, ⟨x, mem_index hs x, rfl⟩⟩
+  have hjk : j = k := hk_unique j ⟨trivial, ⟨y, mem_index hs y, by rw [hfxy, hjy]⟩⟩
+  have hij : i = j := hik.trans hjk.symm
+  rw [hix, hjy, hij] at hfxy
+  refine (h_injOn j) ?_ (mem_index hs y) hfxy
+  rw [← hij]
+  exact mem_index hs x
```
Comment (jcommelin) [inline Mathlib/Data/Setoid/Partition.lean:536]:
```suggestion
  intro x y h
  suffices hs.index x = hs.index y by
    apply h_injOn (hs.index x) (hs.mem_index x) (this ▸ hs.mem_index y)
    simpa only [piecewise_apply, this] using h
  apply h_disjoint.elim trivial trivial
  contrapose! h
  exact h.ne_of_mem (mem_image_of_mem _ (hs.mem_index x)) (mem_image_of_mem _ (hs.mem_index y))
```

File: Mathlib/Data/Setoid/Partition.lean [hunk 6]
```diff
@@ -502,4 +503,58 @@ theorem proj_fiber (x : hs.Quotient) : hs.proj ⁻¹' {x} = s (hs.equivQuotient.
...
+  have hg_bij : ∀ i, BijOn g (s i) (t i) := by
+    intro i
+    refine BijOn.congr (hf i) ?_
+    intro x hx
+    rw [hg, piecewise, hs.mem_iff_index_eq.mp hx]
+  have hg_inj : InjOn g (⋃ i, s i) := by
+    refine injOn_of_injective ?_ (⋃ (i : ι), s i)
+    refine piecewise_inj hs (fun i ↦ BijOn.injOn (hf i)) ?h_disjoint
+    simp only [fun i ↦ BijOn.image_eq (hf i)]
+    intro i _ j _ hij
```
Comment (jcommelin) [inline Mathlib/Data/Setoid/Partition.lean:555]:
```suggestion
    rintro i - j - hij
```

File: Mathlib/Data/Setoid/Partition.lean [hunk 7]
```diff
@@ -502,4 +503,58 @@ theorem proj_fiber (x : hs.Quotient) : hs.proj ⁻¹' {x} = s (hs.equivQuotient.
...
+theorem piecewise_bij {β : Type*} {f : ι → α → β}
+    {t : ι → Set β} (ht : IndexedPartition t)
+    (hf : ∀ i, BijOn (f i) (s i) (t i)) :
+    Bijective (piecewise hs f) := by
+  set g := piecewise hs f with hg
+  have hg_bij : ∀ i, BijOn g (s i) (t i) := by
+    intro i
+    refine BijOn.congr (hf i) ?_
+    intro x hx
+    rw [hg, piecewise, hs.mem_iff_index_eq.mp hx]
```
Comment (jcommelin) [inline Mathlib/Data/Setoid/Partition.lean:550]:
```suggestion
    rw [hg, piecewise_apply, hs.mem_iff_index_eq.mp hx]
```

File: Mathlib/Data/Set/Pairwise/Lattice.lean [hunk 8]
```diff
@@ -164,4 +164,18 @@ theorem Pairwise.biUnion_injective (h₀ : Pairwise (Disjoint on f)) (h₁ : ∀
...
+    (hy : y ∈ (⋃ i ∈ s, f i)) : ∃! i, i ∈ s ∧ y ∈ f i := by
+  refine exists_unique_of_exists_of_unique ?ex ?unique
+  · rw [mem_iUnion] at hy
+    choose i hi using hy
+    use i
+    rw [mem_iUnion] at hi
+    exact exists_prop.mp hi
+  · intro i j hi hj
+    have : ¬Disjoint (f i) (f j) := not_disjoint_iff.mpr ⟨y, ⟨hi.2, hj.2⟩⟩
+    exact PairwiseDisjoint.elim h_disjoint hi.1 hj.1 this
```
Comment (jcommelin) [inline Mathlib/Data/Set/Pairwise/Lattice.lean:179]:
```suggestion
  · simpa only [mem_iUnion, exists_prop] using hy
  · rintro i j ⟨his, hi⟩ ⟨hjs, hj⟩
    exact hsf.elim his hjs <| not_disjoint_iff.mpr ⟨y, ⟨hi, hj⟩⟩
```

**Review by jcommelin — APPROVED** [review]
LGTM, but there are some opportunities to streamline the proofs.

bors d+

File: Mathlib/Data/Setoid/Partition.lean [hunk 9]
```diff
@@ -1,12 +1,13 @@
 /-
 Copyright (c) 2019 Amelia Livingston. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Amelia Livingston, Bryan Gin-ge Chen, Patrick Massot
+Authors: Amelia Livingston, Bryan Gin-ge Chen, Patrick Massot, Wen Yang
```
Comment (shuxuezhuyi) [inline Mathlib/Data/Setoid/Partition.lean:4]:
```suggestion
Authors: Amelia Livingston, Bryan Gin-ge Chen, Patrick Massot, Wen Yang, Johan Commelin
```

**Pull request closed without merging**